fetch/cosmwasm_contract/src/access_control.rs

```diff
@@ -0,0 +1,130 @@
+use std::str::{FromStr, from_utf8};
+use cosmwasm_std::{
+    HumanAddr,
+    StdError,
+    StdResult,
+    Storage,
+    ReadonlyStorage,
+};
+use cosmwasm_storage::{ReadonlyPrefixedStorage, PrefixedStorage};
+
+
+pub static ACCESS_CONTROL_KEY: &[u8] = b"access_control";
+
+pub const OWNER_KEY: &[u8] = b"OWNER_KEY";
+pub const ADMIN_ROLE: &str = "ADMIN_ROLE";
+pub const DELEGATE_ROLE: &str = "DELEGATE_ROLE";
+pub const RELAYER_ROLE: &str = "RELAYER_ROLE";
+
+#[derive(Debug, PartialEq)]
+pub enum AccessRole {
+    Owner,
+    Admin,
+    Delegate,
+    Relayer,
+}
+
+impl AccessRole {
+    fn value(&self) -> &str {
+        match *self {
+            AccessRole::Admin => ADMIN_ROLE,
+            AccessRole::Delegate => DELEGATE_ROLE,
+            AccessRole::Relayer => RELAYER_ROLE,
+            AccessRole::Owner => "",
+        }
+    }
+    
+    fn as_bytes(&self) -> &[u8] {
+        return self.value().as_bytes()
+    }
+    
+    // FIXME(LR) what happen when FromStr treat is in scope?
+    pub fn from_str(s: &str) -> Result<Self, StdError> {
+        match s {
+            ADMIN_ROLE => Ok(AccessRole::Admin),
+            DELEGATE_ROLE => Ok(AccessRole::Delegate),
+            RELAYER_ROLE => Ok(AccessRole::Relayer),
+            _ => Err(StdError::generic_err("Unknow role")),
+        }
+    }
+}
+
+// FIXME(LR) cannot use fn if Trait not in scope, rather annoying
+impl FromStr for AccessRole {
+    type Err = StdError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            ADMIN_ROLE => Ok(AccessRole::Admin),
+            DELEGATE_ROLE => Ok(AccessRole::Delegate),
+            RELAYER_ROLE => Ok(AccessRole::Relayer),
+            _ => Err(StdError::generic_err("Unknow role")),
+        }
+    }
+}
+
+pub fn access_control<S: Storage>(storage: &mut S) -> PrefixedStorage<S> {
+    return PrefixedStorage::new(ACCESS_CONTROL_KEY, storage);
+}
+
+pub fn access_control_read<S: Storage>(storage: &S) -> ReadonlyPrefixedStorage<S> {
+    return ReadonlyPrefixedStorage::new(ACCESS_CONTROL_KEY, storage)
+}
+
+pub fn ac_have_role<S: Storage>(storage: &S, addr: &HumanAddr, role: &AccessRole) -> StdResult<bool> {
+    let ac_store = access_control_read(storage);
+    let addr_roles = ReadonlyPrefixedStorage::new(addr.as_str().as_bytes(), &ac_store);
+    match addr_roles.get(role.as_bytes()) {
+        Some(_) => Ok(true),
+        None => Err(StdError::unauthorized()),
+    }
+}
+
+pub fn ac_add_role<S: Storage>(storage: &mut S, addr: &HumanAddr, role: &AccessRole) -> StdResult<bool> {
```

Comment:
Perhaps I'm missing something, but it feels like anyone can call this function. If my observation is correct, then this needs to be changed to admin only.

Reply:
this is just a helper function *not* a contract call,
even if it is a public function it cannot be called directly on the contract

Reply:
I haven't look in to the higher(caller) layer (the contract in this case), since this component (`access_control.rs` in this case) should be self-containt - should contain the protection logic in.

This is due to reusability reason - if this component is shared/reused by other contract, then first that contract would need to be aware that this check is not in, and second it would need to implementit (again).
This opens potential new contract to unnecessary risk, since developer of the new contract might not be aware of this, assuming it is already in ... 

The method already performs other checks (`if already_have_role {...}`).

I would put this check in to this helper functions to make sure that AccessControl component is self-containt.
Plus, it might be better make these helper functions as member methods of the `AccessRole` type.

Reply:
I understand your point @pbukva and I agree it would be neat to have the check, but I don't think you can *efficiently* implement it at the access_control.rs lib level for many reasons.
> since this component (access_control.rs in this case) should be self-containt - should contain the protection logic in.

it is self-contained, it is an access control store utils not enforcement
> This opens potential new contract to unnecessary risk, since developer of the new contract might not be aware of this, assuming it is already in ...

developers can see from the api that the caller is not passed in, meaning there is no way for the lib to test it.
I think you are projecting the solidity model where the object receiver pointer is implicitly made accessible to function calls

> Another take on this - this helper function already performs the if already_have_role {...} check, then why not the other one?

because it is self-contained

> Plus, it might be better to make all helper functions as member methods of the AccessRole type.
I ma not sure about this

Reply:
The `env: &Env` can be passed in as additional parameter to these functions - the same way as it is done with verification methods `onlyAdmin(...)`, `onlyDelegate(...)`, etc. ... in `contract.rs` module.

I get your point. Perhaps it would be better to take this offline.

Notes: 
> developers can see from the api that the caller is not passed in, meaning there is no way for the lib to do the check.

Bringing attention of developer/programmer to specific aspect works the other way around: If something is necessary to be passed-in explicitly (e.g. `env: &Env` param), then it brings attention to developers by forcing them to do so. If something is simply not there (implicitly *not* necessary to be provided) then it creates potential for simply mis it without even being aware of it (= miss the point).

Reply:
a potential way of achieving it is to raise the abstraction level of the library by moving all access control management in there (i.e., onlyAdmin, onlyDelegate, ...) and add functions with similar signatures for adding/revoking different roles (addAdmin, addDelegate, addRelayer, removeAdmin, removeDelegate, removeRelayer)